File: modules/choice/app/controllers/choice/QuestionController.java
package controllers.choice;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import javax.inject.Inject;

import controllers.core.AbstractController;
import controllers.core.ExerciseController;
import model.Answer;
import model.AnswerKey;
import model.Correctness;
import model.Question;
import model.QuestionRating;
import model.QuestionRatingKey;
import model.QuestionResult;
import model.QuestionType;
import model.QuestionUser;
import model.Quiz;
import model.Util;
import model.user.Role;
import model.user.User;
import play.Logger;
import play.data.DynamicForm;
import play.data.FormFactory;
import play.mvc.Result;

public class QuestionController extends ExerciseController {

  private static final String ATTEMPT_FIELD = "attempt";
  private static final String QUIZ_ID_FIELD = "quizId";

  @Inject
  public QuestionController(Util theUtil, FormFactory theFactory) {
    super(theUtil, theFactory);
  }

  public static User getUser() {
    User user = ExerciseController.getUser();

    if(QuestionUser.finder.byId(user.name) == null)
      // Make sure there is a corresponding entrance in other db...
      new QuestionUser(user.name).save();

    return user;
  }

  private static void saveRating(String name, int id, int stars) {
    QuestionRatingKey key = new QuestionRatingKey(id, name);
    QuestionRating rating = QuestionRating.finder.byId(key);

    if(rating == null)
      rating = new QuestionRating(key);

    rating.rating = stars;
    rating.save();
  }

  public Result correct(int id) {
    User user = getUser();

    Question question = Question.finder.byId(id);
    DynamicForm form = factory.form().bindFromRequest();

    List<Answer> selectedAnswers = readSelAnswers(question, form);
    QuestionResult result = new QuestionResult(selectedAnswers, question);

    int stars = Integer.parseInt(form.get("stars"));
    if(stars != -1)
      saveRating(user.name, question.id, stars);

    return ok(views.html.questionResult.render(user, result));
  }

  public Result editQuestion(int id) {
    DynamicForm form = factory.form().bindFromRequest();

    Logger.debug("DATA:\n" + form.data());

    Question question = Question.finder.byId(id);

    question.title = form.get("title");
    question.text = form.get("text");
    question.author = getUser().name;
    question.questionType = QuestionType.valueOf(form.get("type"));

    int numOfAnswers = Integer.parseInt(form.get("numOfAnswers"));
    question.answers = new ArrayList<>(numOfAnswers);

    for(int count = 1; count <= numOfAnswers; count++) {
      AnswerKey key = new AnswerKey(id, count);
      Answer answer = Answer.finder.byId(key);
      if(answer == null)
        answer = new Answer(key);
      answer.correctness = Correctness.valueOf(form.get("correctness_" + count));
      answer.text = form.get(Integer.toString(count));

      question.answers.add(answer);
    }

    question.save();
    for(Answer answer: question.answers)
      answer.save();

    return ok(views.html.questionadmin.choiceCreation.render(getUser(), Arrays.asList(question)));
  }

  public Result editQuestionForm(int id) {
    User user = getUser();
    Question question = Question.finder.byId(id);

    if(question.author == user.name || user.role == Role.ADMIN)
      return ok(views.html.editQuestionForm.render(user, question));

    return redirect(routes.QuestionController.index());
  }

  public Result index() {
    return ok(views.html.choiceoverview.render(getUser(), Quiz.finder.all()));
  }

  public Result newQuestion() {
    DynamicForm form = factory.form().bindFromRequest();

    int id = findMinimalNotUsedId(Question.finder);

    Question question = new Question(id);
    question.title = form.get("title");
    question.text = form.get("text");
    question.author = getUser().name;
    question.questionType = QuestionType.valueOf(form.get("type"));

    // FIXME: Answers...
    int numOfAnswers = Integer.parseInt(form.get("numOfAnswers"));
    question.answers = new ArrayList<>(numOfAnswers);

    for(int count = 1; count <= numOfAnswers; count++) {
      AnswerKey key = new AnswerKey(id, count);
      Answer answer = Answer.finder.byId(key);
      if(answer == null)
        answer = new Answer(key);
      // TODO: Korrektheit der Antwort bei MC und SC!
      answer.correctness = Correctness.WRONG; // Correctness.valueOf("");
      answer.text = form.get(Integer.toString(count));

      question.answers.add(answer);
    }

    question.save();
    for(Answer answer: question.answers)
      answer.save();

    return ok(views.html.questionadmin.choiceCreation.render(getUser(), Arrays.asList(question)));
  }

  public Result newQuestionForm() {
    return ok(views.html.newquestionform.render(getUser()));
  }

  public Result question(int id) {
    Question question = Question.finder.byId(id);

    if(question == null)
      return redirect(controllers.choice.routes.QuestionController.index());

    switch(question.questionType) {
    case MULTIPLE:
    case SINGLE:
      return ok(views.html.choiceQuestion.render(getUser(), question));
    case FILLOUT_WITH_ORDER:
    case FILLOUT_WITHOUT_ORDER:
      return ok(views.html.filloutquestion.render(getUser(), question));
    default:
      return redirect(routes.QuestionController.index());
    }
  }

  public Result questions() {
    return ok(views.html.questions.render(getUser(), Question.finder.all()));
  }

  public Result quiz(int id) {
    return ok(views.html.quiz.render(getUser(), Quiz.finder.byId(id)));
  }

  public Result quizCorrection(int quizId, int questionId) {
    User user = getUser();

    Quiz quiz = Quiz.finder.byId(quizId);

    Question question = quiz.questions.get(questionId - 1);
    DynamicForm form = factory.form().bindFromRequest();

    List<Answer> selectedAnswers = readSelAnswers(question, form);
    QuestionResult result = new QuestionResult(selectedAnswers, question);

    int stars = Integer.parseInt(form.get("stars"));
    if(stars != -1)
      saveRating(user.name, question.id, stars);

    return ok(views.html.quizQuestionResult.render(user, quiz, result));
  }

  public Result quizQuestion(int quizId, int questionId) {
    Logger.debug("Attempt: " + session().get(ATTEMPT_FIELD));
    return ok(views.html.quizQuestion.render(getUser(), Quiz.finder.byId(quizId), questionId - 1));
  }

  public Result quizStart(int quizId) {
    // TODO: from db...
    int attempt = 1;

    // TODO: Initialize session
    session().put(ATTEMPT_FIELD, Integer.toString(attempt));
    session().put(QUIZ_ID_FIELD, Integer.toString(quizId));

    return redirect(routes.QuestionController.quizQuestion(quizId, 1));
  }

  public Result quizzes() {
    return ok(views.html.quizzes.render(getUser(), Quiz.finder.all()));
  }

  private List<Answer> readSelAnswers(Question question, DynamicForm form) {
    Logger.debug(question.questionType.toString());
    switch(question.questionType) {
    case MULTIPLE:
    // @formatter:off
    return question.answers.stream()
        .filter(ans -> form.get(Integer.toString(ans.key.id)) != null)
        .collect(Collectors.toList());
    //@formatter:on
    case SINGLE:
      return new ArrayList<>(Arrays.asList(question.getAnswer(Integer.parseInt(form.get("ans")))));
    case FILLOUT_WITH_ORDER:
    case FILLOUT_WITHOUT_ORDER:
    default:
      return Collections.emptyList();
    }
  }

}

File: app/controllers/exCollections/QuestionController.scala
package controllers.exCollections

import javax.inject.{Inject, Singleton}

import controllers.Secured
import model.Enums.Role
import model.core._
import model.questions.QuestionEnums.QuestionType
import model.questions._
import model.{JsonFormat, User}
import play.api.db.slick.{DatabaseConfigProvider, HasDatabaseConfigProvider}
import play.api.libs.json._
import play.api.mvc._
import play.twirl.api.Html
import slick.jdbc.JdbcProfile

import scala.concurrent.{ExecutionContext, Future}
import scala.language.{implicitConversions, postfixOps}
import scala.util.{Failure, Try}

@Singleton
class QuestionController @Inject()(cc: ControllerComponents, dbcp: DatabaseConfigProvider, r: Repository)(implicit ec: ExecutionContext)
  extends AExCollectionController[Question, Quiz, IdAnswerMatch, QuestionResult](cc, dbcp, r, QuestionToolObject)
    with HasDatabaseConfigProvider[JdbcProfile] with JsonFormat with Secured {

  override type SolType = Seq[GivenAnswer]

  override def readSolutionFromPutRequest(implicit request: Request[AnyContent]): Option[Seq[GivenAnswer]] = request.body.asJson flatMap (_.asObj) flatMap { jsObj =>
    jsObj.stringField("questionType") flatMap QuestionType.byString flatMap {
      case QuestionType.CHOICE   => jsObj.arrayField("chosen", jsValue => Some(IdGivenAnswer(jsValue.asInt getOrElse -1)))
      case QuestionType.FREETEXT => ??? // Some(Seq.empty)
    }
  }

  override def readSolutionFromPostRequest(implicit request: Request[AnyContent]): Option[Seq[GivenAnswer]] = {
    println(request.body.asFormUrlEncoded)

    None
  }

  // Yaml

  override type CompColl = CompleteQuiz

  override type CompEx = CompleteQuestion

  override implicit val yamlFormat: net.jcazevedo.moultingyaml.YamlFormat[CompleteQuiz] = QuestionYamlProtocol.QuizYamlFormat

  // db

  override type TQ = repo.QuizzesTable

  override def tq = repo.quizzes

  override protected def numOfExesInColl(id: Int): Future[Int] = repo.questionsInQuiz(id)

  override protected def futureCompleteColls: Future[Seq[CompColl]] = repo.completeQuizzes

  override protected def futureCompleteCollById(id: Int): Future[Option[CompleteQuiz]] = repo.completeQuiz(id)

  override protected def futureCompleteExById(collId: Int, id: Int): Future[Option[CompleteQuestion]] = repo.completeQuestion(collId, id)

  override protected def saveRead(read: Seq[CompColl]): Future[Seq[Boolean]] = Future.sequence(read map repo.saveQuiz)

  // Other routes

  def quizStart(quizId: Int): EssentialAction = withUser { _ =>
    implicit request => Redirect(routes.QuestionController.exercise(quizId, 1))
  }

  //  def quizzes: EssentialAction = withUser { user => implicit request => Ok(views.html.questions.quizzes.render(user, null /*Quiz.finder.all.asScala.toList*/)) }

  def assignQuestion(keyAndValue: String, addOrRemove: Boolean) {
    // String[] quizAndQuestion = keyAndValue.split("_")
    //
    // Quiz quiz = Quiz.finder.byId(Integer.parseInt(quizAndQuestion[0]))
    // Question question =
    // Question.finder.byId(Integer.parseInt(quizAndQuestion[1]))
    //
    // if(addOrRemove)
    // quiz.questions.add(question)
    // else
    // quiz.questions.remove(question)
    //
    // quiz.save()
  }

  def assignQuestions: EssentialAction = withAdmin { _ =>
    implicit request =>
      //    val form = factory.form().bindFromRequest()

      // Read it...
      //    Map<String, String> assignments = form.rawData()
      //    for(Map.Entry<String, String> entry: assignments.entrySet())
      //      assignQuestion(entry.getKey(), "on".equals(entry.getValue()))
      //
      //    return ok(views.html.questionAdmin.questionsAssigned.render(user(), assignments.toString()))
      Ok("TODO!")
  }

  def assignQuestionsForm: EssentialAction = withAdmin { user =>
    implicit request =>
      Ok(views.html.questions.questionAdmin.assignQuestionsForm.render(user, /* Question.finder.all() */ List.empty, List.empty /*Quiz.finder.all.asScala.toList*/))
  }

  def assignQuestionsSingleForm(id: Int): EssentialAction = withAdmin { user =>
    implicit request =>
      Ok(views.html.questions.questionAdmin.assignQuestionsForm.render(user, /* Question.finder.all() */ List.empty, List.empty /*(Quiz.finder.byId(id))*/))
  }

  //  def  exportQuizzes = {
  //    val json = Json.prettyPrint(Json.toJson(Quiz.finder.all()))
  //
  //    try {
  //      File tempFile = new File("quizzes_export_" + LocalDateTime.now() + ".json")
  //      Files.asCharSink(tempFile, Charset.defaultCharset()).write(json)
  //      // false == download file!
  //      Ok(tempFile, false)
  //    } catch (IOException e) {
  //      Ok(json)
  //    }
  //  }

  def gradeFreetextAnswer(id: Int, user: String): EssentialAction = withAdmin { _ =>
    implicit request =>
      // FreetextAnswer answer = FreetextAnswer.finder.byId(new
      // FreetextAnswerKey(user, id))
      //             views.html.questionAdmin.ftaGradeForm.render(user(), answer)
      Ok("TODO")
  }

  def gradeFreetextAnswers: EssentialAction = withAdmin { _ => implicit request => Ok("TODO!") }

  // views.html.questionAdmin.ftasToGrade.render(user(), FreetextAnswer.finder.all())

  def notAssignedQuestions: EssentialAction = withAdmin { user =>
    implicit request =>
      Ok(views.html.questions.questionList.render(user, List.empty /* Question.notAssignedQuestions() */))
  }

  // User

  // private static List<Answer> readAnswersFromForm(DynamicForm form, int
  // questionId, boolean isChoice) {
  // return IntStream.range(0, Question.MAX_ANSWERS).mapToObj(id -> {
  // final AnswerKey key = new AnswerKey(questionId, id)
  //
  // Answer answer = Answer.finder.byId(key)
  // if(answer == null)
  // answer = new Answer(key)
  //
  // answer.text = form.get(String.valueOf(id))
  // answer.correctness = isChoice ? Correctness.valueOf(form.get("correctness_"
  // + id)) : Correctness.CORRECT
  //
  // return answer
  // }).filter(ans -> !ans.text.isEmpty()).collect(Collectors.toList())
  // }

  //  private def readSelAnswers(question: Question, form: DynamicForm): List[Answer] =
  //    question.answers.asScala.filter(ans => Option(form.get(ans.key.id.toString)).isDefined).toList

  //  override def correctEx(sol: StringSolution, exercise: Option[Question], user: User): Try[CompleteResult[QuestionResult]] = ???

  def editQuestion(id: Int): EssentialAction = withUser { user =>
    implicit request =>
      // DynamicForm form = factory.form().bindFromRequest()

      val question: Question = null // readQuestionFromForm(form,
      // Question.finder.byId(id))
      // question.save()
      // for(Answer answer: question.answers)
      // answer.save()

      Ok(views.html.questions.questionAdmin.questionCreated.render(user, List(question)))
  }

  def editQuestionForm(id: Int): EssentialAction = futureWithUser { user =>
    implicit request =>
      futureCompleteExById(1 /*collId*/ , id) map {
        case None           => BadRequest("TODO!")
        case Some(question) =>
          if (question.author == user.username || user.stdRole == Role.RoleAdmin)
            Ok(views.html.questions.editQuestionForm.render(user, question, isEdit = true))
          else
            Redirect(routes.QuestionController.index())
      }
  }

  def newQuestion(isFreetext: Boolean): EssentialAction = withUser { _ =>
    implicit request =>
      //    QuestionReader.initFromForm(0, null /* factory.form().bindFromRequest()*/) match {
      //      case ReadingError(_, _, _) => BadRequest("There has been an error...")
      //      case ReadingFailure(_) => BadRequest("There has been an error")
      //      case ReadingResult(questions) =>
      //        val question = questions.head.read.asInstanceOf[Question]
      //        QuestionReader.save(question)
      //        Ok(views.html.questionAdmin.questionCreated.render(user, List(question)))
      //    }
      Ok("TODO!")
  }

  def newQuestionForm(isFreetext: Boolean): EssentialAction = withUser { user =>
    implicit request =>
      if (isFreetext) Ok(views.html.questions.newFreetextQuestionForm.render(user, Integer.MAX_VALUE))
      else {
        // TODO: Unterscheidung zwischen ausfuellen und ankreuzen!
        Ok(views.html.questions.newQuestionForm.render(user, Integer.MAX_VALUE, isChoice = true))
      }
  }

  //  def questionResult(id: Int): EssentialAction = withUser { user =>
  //    implicit request =>
  //      val question = Question.finder.byId(id)
  //
  //      if (question.questionType != Question.QType.FREETEXT) {
  //        // FILLOUT or MULTIPLE CHOICE
  //        //      val form = factory.form().bindFromRequest()
  //        //      val result = new QuestionResult(readSelAnswers(question, form), question)
  //        Ok(views.html.givenanswerQuestionResult.render(user, null /* result*/))
  //      } else {
  //        val key = new UserAnswerKey(user.name, id)
  //        val answer = Option(UserAnswer.finder.byId(key)).getOrElse(new UserAnswer(key))
  //
  //
  //        answer.question = Question.finder.byId(id)
  //        //    answer.text = factory.form().bindFromRequest().get("answer")
  //        answer.save()
  //
  //        Ok(views.html.freetextQuestionResult.render(user, question, answer))
  //  }
  //      Ok("TODO!")
  //  }

  // Views

  override def adminRenderEditRest(collOpt: Option[CompleteQuiz]): Html = new Html(
    s"""<div class="form-group row">
       |  <div class="col-sm-12">
       |    <label for="${QuestionConsts.ThemeName}">Thema:</label>
       |    <input class="form-control" name="${QuestionConsts.ThemeName}" id="${QuestionConsts.ThemeName}" required ${collOpt map (coll => s"""value="${coll.coll.theme}"""") getOrElse ""})>
       |  </div>
       |</div>""".stripMargin)

  override protected def renderExercise(user: User, quiz: Quiz, exercise: CompleteQuestion, numOfExes: Int): Html =
    views.html.questions.question(user, quiz, exercise, numOfExes, None /* FIXME: old answer... UserAnswer.finder.byId(new UserAnswerKey(user.name, exercise.id))*/)

  protected def onSubmitCorrectionError(user: User, error: Throwable): Result = ???

  protected def onSubmitCorrectionResult(user: User, result: QuestionResult): Result = ???

  protected def onLiveCorrectionError(error: Throwable): Result = ???

  protected def onLiveCorrectionResult(result: QuestionResult): Result = Ok(Json.obj(
    "correct" -> JsArray(result.correct map (i => JsNumber(i.id))),
    "missing" -> JsArray(result.missing map (i => JsNumber(i.id))),
    "wrong" -> JsArray(result.wrong map (i => JsNumber(i.id)))
  ))

  // Correction

  override def correctEx(user: User, answers: Seq[GivenAnswer], exercise: CompleteQuestion, quiz: Quiz): Try[QuestionResult] = exercise.ex.questionType match {
    case QuestionType.FREETEXT => Failure(new Exception("Not yet implemented..."))
    case QuestionType.CHOICE   => Try {
      val idAnswers: Seq[IdGivenAnswer] = answers flatMap {
        case idA: IdGivenAnswer => Some(idA)
        case _                  => None
      }

      QuestionResult(idAnswers, exercise)
      // ???
    }
  }
}
